Add Netty Based HTTP2 implementation for storage server.

zzmao opened the pull request:
Add AsyncWritableChannel interface for MessageReadSet and Send.
Add StorageServerNettyFactory and StorageServerNettyChannelInitializer.
Add HTTP2 availability in AmbryServer but it is disabled unitl HTTP2 port populated in Helix.
Add ServerHttp2Test.

zzmao commented on ambry-rest/src/test/java/com.github.ambry.rest/AsyncRequestResponseHandlerFactoryTest.java:
```diff
@@ -71,7 +71,6 @@ public void getAsyncRequestResponseHandlerTest() throws InstantiationException,
   /**
    * Tests instantiation of {@link AsyncRequestResponseHandlerFactory} with bad input.
    */
-  @Test
```
The test is temporarily disabled, will fix it soon in this PR.

zzmao replied:
fixed.

zzmao commented on ambry-rest/src/main/java/com.github.ambry.rest/AsyncRequestResponseHandlerFactory.java:
```diff
@@ -109,10 +109,10 @@ private static void buildInstance(MetricRegistry metricRegistry) {
         instance = new AsyncRequestResponseHandler(requestResponseHandlerMetrics);
       }
       // check if same instance of MetricRegistry - otherwise it is a problem.
-      if (AsyncRequestResponseHandlerFactory.requestResponseHandlerMetrics.metricRegistry != metricRegistry) {
```
Temporarily commented. Will fix it soon in this PR.

zzmao replied:
fixed.

justinlin-linkedin commented on ambry-protocol/src/main/java/com.github.ambry.protocol/CompositeSend.java:
```diff
@@ -48,6 +50,13 @@ public long writeTo(WritableByteChannel channel) throws IOException {
+      send.writeTo(channel, callback);
+    }
+  }
+
```
we should only call callback method once for the entire list since the callback passed to writeTo should be invoked only when the write is completed or failed.

cgtz replied:
Yes, perhaps each child send should have its own callback and then when either all of those callbacks succeed or an error has occured, the passed-in callback is notified, like in the default impl of `AsyncWritableChannel.write(ByteBuf src, Callback<Long> callback)`

zzmao replied:
Yes, this should be changed. 
I will make another PR to address this issue. For this PR, I will set callback to null in child send.

justinlin-linkedin commented on ambry-protocol/src/main/java/com.github.ambry.protocol/GetResponse.java:
```diff
@@ -131,6 +133,26 @@ public long writeTo(WritableByteChannel channel) throws IOException {
+    channel.write(bufferToSend, callback);
+    if (toSend != null) {
+      toSend.writeTo(channel, callback);
+    }
```
same here, callback should only be invoked once.

zzmao replied:
Will make another PR to address this issue. For this PR, I will set callback to null in child send.

cgtz commented on ambry-api/src/main/java/com.github.ambry/store/MessageReadSet.java:
```diff
@@ -36,7 +38,14 @@
 
   /**
-   * Returns the total number of messages in this set
+   * Write all messages in this set to the give {@link AsyncWritableChannel}
```
Add asynchronously to the docs to indicate that this method returns immediately after the write is initiated?

zzmao replied:
Added in comment:
```
   * This method is intend to write prefetched data from {@link MessageReadSet} to {@link AsyncWritableChannel}. Data
   * should be ready in memory(no blocking call) before write to {@link AsyncWritableChannel} asynchronously. Callback is
   * called when the entire batch of writes succeeds or fails.
```

cgtz commented on ambry-api/src/main/java/com.github.ambry/store/MessageReadSet.java:
```diff
@@ -36,7 +38,14 @@
+   * @param callback The callback when data is fully wrote to the channel.
+   * @throws IOException
+   */
+  void writeTo(AsyncWritableChannel channel, Callback callback) throws IOException;
```
`Callback<Void>`?

cgtz commented on ambry-protocol/src/main/java/com.github.ambry.protocol/CompositeSend.java:
```diff
@@ -48,6 +50,13 @@ public long writeTo(WritableByteChannel channel) throws IOException {
 
+  @Override
+  public void writeTo(AsyncWritableChannel channel, Callback callback) throws IOException {
+    for (Send send : compositSendList) {
```
composit -> composite

cgtz commented on ambry-network/src/main/java/com.github.ambry.network/NettyServerRequestResponseChannel.java:
```diff
@@ -32,6 +33,13 @@ public NettyServerRequestResponseChannel(int queueSize) {
   public void sendRequest(NetworkRequest request) throws InterruptedException {
+    DataInputStream stream = new DataInputStream(request.getInputStream());
+    try {
+      // TODO: Where is this done in socket channel?
```
This is done inside of the selector, reading the size of the payload is implemented in `BoundedReceive`

cgtz commented on ambry-rest/src/main/java/com.github.ambry.rest/StorageServerNettyFactory.java:
```diff
@@ -0,0 +1,88 @@
+   * @param publicAccessLogger the {@link PublicAccessLogger} that can be used for public access logging
+   * @param restServerState the {@link RestServerState} that can be used to check the health of the system
+   *                              to respond to health check requests
+   * @param defaultSslFactory the {@link SSLFactory} used to construct the {@link javax.net.ssl.SSLEngine} used for
```
It does not look like this factory has the override behavior, so you can modify this javadoc and change the arg name to just `sslFactory`

cgtz commented on ambry-rest/src/main/java/com.github.ambry.rest/StorageServerNettyFactory.java:
```diff
@@ -0,0 +1,88 @@
+      final PublicAccessLogger publicAccessLogger, final RestServerState restServerState, SSLFactory defaultSslFactory)
+      throws ReflectiveOperationException {
+    if (verifiableProperties == null || metricRegistry == null || requestHandler == null || publicAccessLogger == null
+        || restServerState == null) {
```
sslfactory is required

cgtz commented on ambry-rest/src/main/java/com.github.ambry.rest/StorageServerNettyFactory.java:
```diff
@@ -0,0 +1,88 @@
+
+    initializers.put(http2Port,
+        new StorageServerNettyChannelInitializer(nettyConfig, performanceConfig, nettyMetrics, connectionStatsHandler,
+            requestHandler, publicAccessLogger, restServerState, defaultSslFactory, metricRegistry));
```
It may be worth looking into what to do about public access logging on storage servers. They already have public access logging facilities inside of AmbryRequests. We would either want to disable these logs or make sure that they are redirected to the right log files.

zzmao replied:
removed.

cgtz commented on ambry-rest/src/main/java/com.github.ambry.rest/StorageServerNettyFactory.java:
```diff
@@ -0,0 +1,88 @@
+    nettyMetrics = new NettyMetrics(metricRegistry);
+    ConnectionStatsHandler connectionStatsHandler = new ConnectionStatsHandler(nettyMetrics);
+
+    Map<Integer, ChannelInitializer<SocketChannel>> initializers = new HashMap<>();
```
you can just use `Collections.singletonMap` here.

zzmao replied:
cool.

cgtz commented on ambry-rest/src/main/java/com.github.ambry.rest/StorageServerNettyChannelInitializer.java:
```diff
@@ -0,0 +1,95 @@
+    }
+    // TODO: add PublicAccessLogger and ConnectionStatsHandler
+    pipeline.addLast(Http2FrameCodecBuilder.forServer().build())
+        // .addLast("healthCheckHandler", new HealthCheckHandler(restServerState, nettyMetrics))
```
commented out ^

zzmao replied:
removed.

cgtz commented on ambry-rest/src/main/java/com.github.ambry.rest/Http2StreamHandler.java:
```diff
@@ -0,0 +1,51 @@
+  @Override
+  public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+    ctx.pipeline().addLast(new Http2StreamFrameToHttpObjectCodec(true));
+    // NettyMessageProcessor depends on ChunkedWriteHandler.
```
Do the utility handlers like HealthCheckHandler, and PublicAccessLogHandler belong here?

zzmao replied:
Existing server logging is good. We don't need PublicAccessLogHandler.
`HealthCheckHandler ` is needed. Let me if I can add it in this PR.

zzmao commented on ambry-protocol/src/main/java/com.github.ambry.protocol/GetResponse.java:
```diff
@@ -131,6 +133,26 @@ public long writeTo(WritableByteChannel channel) throws IOException {
+        }
+      }
+      bufferToSend.flip();
+    }
```
yes, I should do this.

zzmao commented:
Addressed most of the comments(except for those explicitly mentioned in comments). Please take another round of review. @cgtz @justinlin-linkedin Thanks!

justinlin-linkedin commented on ambry-rest/src/main/java/com.github.ambry.rest/StorageServerNettyChannelInitializer.java:
```diff
@@ -0,0 +1,89 @@
+    pipeline.addLast(Http2FrameCodecBuilder.forServer().build())
+        .addLast("IdleStateHandler", new IdleStateHandler(0, 0, nettyConfig.nettyServerIdleTimeSeconds))
+        .addLast("Http2MultiplexHandler", new Http2MultiplexHandler(
+            new Http2StreamHandler(nettyMetrics, nettyConfig, performanceConfig, requestHandler)));
```
I think we should probably add another handler between IdleStateHandler and Http2MultiplexHandler to deal with IdleStateEvent.

Http2MulitiplexHandler doesn't seem like it will handle IdleStateEvent, so we would probably see all the children channels would be triggered by this event.

zzmao replied:
Got you. Let me think about this.

zzmao replied:
OK to remove IdleStateHandler as discussed offline.

zzmao commented on ambry-store/src/main/java/com.github.ambry.store/StoreMessageReadSet.java:
```diff
@@ -208,6 +210,25 @@ public long writeTo(int index, WritableByteChannel channel, long relativeOffset,
+        throw new IOException("Data should be prefetched.");
+      }
+      buf.position(0);
+      if (i == lastIndex) {
```
good catch.

zzmao commented on ambry-store/src/main/java/com.github.ambry.store/StoreMessageReadSet.java:
```diff
@@ -208,6 +210,25 @@ public long writeTo(int index, WritableByteChannel channel, long relativeOffset,
+        channel.write(buf, callback);
+      } else {
+        channel.write(buf, null);
+      }
```
Yes. A good callback implementation is needed. Will add todo here and implement in follow up PR.

zzmao commented on ambry-server/src/test/java/com.github.ambry.server/Http2BlockingChannel.java:
```diff
@@ -103,24 +104,31 @@ public void send(Send request) throws IOException {
-    Http2ResponseHandler.StreamResult streamResult = http2ResponseHandler.awaitResponses(5, TimeUnit.SECONDS);
-    DataInputStream dataInputStream = new NettyByteBufDataInputStream(streamResult.getByteBuf());
+    channelPromise.awaitUninterruptibly(3, TimeUnit.SECONDS);
+    DataInputStream dataInputStream = new NettyByteBufDataInputStream(responseByteBuf);
```
Adding a check if return value T or F.

zzmao commented on ambry-server/src/main/java/com.github.ambry.server/AmbryServer.java:
```diff
@@ -201,6 +214,36 @@ public void startup() throws InstantiationException {
+            new AmbryServerRequests(storageManager, requestResponseChannel, clusterMap, nodeId, registry, serverMetrics,
+                findTokenHelper, notificationSystem, replicationManager, storeKeyFactory, serverConfig,
+                storeKeyConverterFactory, statsManager);
+        RequestHandlerPool requestHandlerPoolForHttp2 =
```
fixed. Also renamed some variables name with `http2` keyword.

zzmao commented on ambry-protocol/src/main/java/com.github.ambry.protocol/AdminResponse.java:
```diff
@@ -59,15 +61,25 @@ public static AdminResponse readFrom(DataInputStream stream) throws IOException
 
-  @Override
-  public long writeTo(WritableByteChannel channel) throws IOException {
+  private void prepareBufferToSend() {
```
added. Also changed in GetResponse.

zzmao commented on ambry-protocol/src/main/java/com.github.ambry.protocol/CompositeSend.java:
```diff
@@ -24,33 +26,48 @@
+        send.writeTo(channel, null);
+      }
+      i++;
+    }
```
What's your suggestion on how to stop it?

zzmao replied:
Had offline chat. Looks like there is no immediate way to stop the loop, because exception happens in callback.  But will enhance callback logic in next PR.

zzmao commented on ambry-network/src/main/java/com.github.ambry.network/NettyServerRequestResponseChannel.java:
```diff
@@ -32,6 +33,14 @@ public NettyServerRequestResponseChannel(int queueSize) {
+    try {
+      // The first 8 bytes is size of the request. TCP implementation uses this size to allocate buffer. See {@link BoundedReceive}
+      // Here we just need to consume it.
+      stream.readLong();
```
Talk offline.

cgtz commented on ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerPlaintextTest.java:
```diff
@@ -74,7 +74,13 @@ public RouterServerPlaintextTest(boolean testEncryption) {
-    plaintextCluster = new MockCluster(new Properties(), false, SystemTime.getInstance());
+
+    Properties serverProperties = new Properties();
+    serverProperties.setProperty("rest.server.rest.request.service.factory",
```
call `TestSSLUtils.addHttp2Properties` here?

zzmao replied:
fixed.

cgtz commented on ambry-server/src/test/java/com.github.ambry.server/Http2BlockingChannel.java:
```diff
@@ -96,31 +103,39 @@ public void disconnect() throws IOException {
 
   @Override
   public void send(Send request) throws IOException {
+    System.out.println("send");
```
remove println

cgtz commented on ambry-server/src/test/java/com.github.ambry.server/Http2BlockingChannel.java:
```diff
@@ -96,31 +103,39 @@ public void disconnect() throws IOException {
+    Http2StreamChannel childChannel = http2StreamChannelBootstrap.open().syncUninterruptibly().getNow();
     Http2Headers http2Headers = new DefaultHttp2Headers().method(HttpMethod.POST.asciiName()).scheme("https").path("/");
+    http2Headers.set(RestUtils.Headers.HTTP2_FRONTEND_REQUEST, "true");
+    channelPromise = childChannel.newPromise();
```
Instead of combining ambry callbacks and netty promises here, how about setting the attribute to a `CompletableFuture` or ambry's `FutureResult`? Then you can wait on the future without requiring lines 135-137

zzmao commented on ambry-api/src/main/java/com.github.ambry/rest/RestUtils.java:
```diff
@@ -216,6 +216,11 @@
+    /**
+     * Request header indicating a HTTP2 request from frontend to server.
+     */
+    public final static String HTTP2_FRONTEND_REQUEST = "x-ambry-http2-frontend-request";
```
how about `x-ambry-internal-http2-request`?

zzmao commented on ambry-protocol/src/main/java/com.github.ambry.protocol/CompositeSend.java:
```diff
@@ -24,33 +26,49 @@
   }
 
+  @Override
+  public void writeTo(AsyncWritableChannel channel, Callback callback) throws IOException {
```
addressing all callback<Long> issue.

zzmao commented on ambry-protocol/src/main/java/com.github.ambry.protocol/Response.java:
```diff
@@ -42,16 +44,26 @@ protected void writeHeader() {
   }
 
+  @Override
+  public void writeTo(AsyncWritableChannel channel, Callback<Long> callback) throws IOException {
```
This is a good suggestion. `IOException` is not needed.

zzmao commented on ambry-rest/src/main/java/com.github.ambry.rest/NettySslHttp2Factory.java:
```diff
@@ -39,7 +39,6 @@
 
 /**
  * An HTTP/2 specific implementation of {@link SSLFactory} that uses Netty's SSL libraries for HTTP2.
```
added back.

zzmao commented on ambry-rest/src/main/java/com.github.ambry.rest/StorageServerNettyChannelInitializer.java:
```diff
@@ -0,0 +1,89 @@
+    ChannelPipeline pipeline = ch.pipeline();
+    // connection stats handler to track connection related metrics
+    // if SSL is enabled, add an SslHandler before the HTTP codec
+    if (sslFactory != null) {
```
added.

zzmao commented on ambry-rest/src/main/java/com.github.ambry.rest/StorageServerNettyChannelInitializer.java:
```diff
@@ -0,0 +1,89 @@
+    // in the pipeline for every connection.
+    // i.e. if there are a 1000 active connections there will be a 1000 NettyMessageProcessor instances.
+    ChannelPipeline pipeline = ch.pipeline();
+    // connection stats handler to track connection related metrics
```
added.

zzmao commented on ambry-network/src/main/java/com.github.ambry.network/NettyServerRequestResponseChannel.java:
```diff
@@ -32,6 +33,14 @@ public NettyServerRequestResponseChannel(int queueSize) {
+      // Here we just need to consume it.
+      stream.readLong();
+    } catch (IOException e) {
+      throw new InterruptedException("stream read error." + e);
```
fixed.

zzmao commented on ambry-protocol/src/main/java/com.github.ambry.protocol/CompositeSend.java:
```diff
@@ -24,33 +26,49 @@
+    for (Send send : compositeSendList) {
+      if (i == lastIndex) {
+        // only the last one pass in callback
+        send.writeTo(channel, callback);
```
added a comment

zzmao commented:
@cgtz  Ready for merge.
